### PR TITLE
Avoid pickled tasty for some captured quote reference

### PR DESCRIPTION
**Pull request by nicolasstucki**
This avoids an extra pickle for `'y` in `'{ val y = $x * $x; ${ powerCode('y, n / 2) } }`.

### CODE REVIEW

**Comment by nicolasstucki** [comment]
test performance with #quotes please

**Comment by nicolasstucki** [comment]
test performance with #quotes please

**Comment by nicolasstucki** [comment]
test performance with #quotes please

**Comment by nicolasstucki** [comment]
@liufengyun what failed in the benchmarks? I cannot access the logs.

**Comment by liufengyun** [comment]
From the error logs in the job processor, I can see is the following:

```
curl: (28) Failed to connect to api.github.com port 443: Connection timed out
```

I checked the file `bin/pull`, the following line failed:

```
curl "https://api.github.com/repos/lampepfl/dotty/pulls/$PR" > $output
```

The line is at the beginning of the PR test script to get PR information.

**Comment by liufengyun** [comment]
test performance with #quotes please

File: compiler/src/dotty/tools/dotc/transform/PickleQuotes.scala [hunk 1]
```diff
@@ -481,10 +481,16 @@ class PickleQuotes extends MacroTransform {
...
+          case Apply(TypeApply(fn, (body: RefTree) :: Nil), _)
+          if fn.symbol == defn.QuotedTypeModule_of && isCaptured(body.symbol, level + 1) =>
             // Optimization: avoid the full conversion when capturing `X` with `x$1: Type[X$1]`
             // in `Type.of[X]` to `Type.of[x$1.Underlying]` and go directly to `X$1`
             capturers(body.symbol)(body)
+          case Apply(Select(Apply(TypeApply(fn,_), List(ref: RefTree)),nme.apply),List(quotes))
+          if fn.symbol == defn.QuotedRuntime_exprQuote && isCaptured(ref.symbol, level + 1) =>
+            // Optimization: avoid the full conversion when capturing `x` with `x$1: Expr[X]`
+            // in `'{x}` to `'{ ${x$1} }'` and go directly to `x$1`
+            capturers(ref.symbol)(ref).select(nme.apply).appliedTo(quotes)
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/transform/PickleQuotes.scala:493]: For the example `'{ val y = $x * $x; ${ powerCode('y, n / 2) } }`, I'm wondering what's the argument to `powerCode` before and after the change.

**Review by liufengyun — APPROVED** [review]
Otherwise, LGTM

File: compiler/src/dotty/tools/dotc/transform/PickleQuotes.scala (hunk 1 above)
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/transform/PickleQuotes.scala:493]: That was the test case I used to do this change. Unless the version of the benchmark has something else that I missed.

**Comment by nicolasstucki** [comment]
test performance with #quotes please

File: compiler/src/dotty/tools/dotc/transform/PickleQuotes.scala (hunk 1 above)
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/transform/PickleQuotes.scala:493]: Before we had

```scala
def hole1(args: Array[Any]) = 
  val y$1 = args(0).asInstanceOf[Expr[Int]]
  def hole3(args: Array[Any]) = y$1
  powerCode(unpickle('{ [[ 0 | Double ]] }, List(hole3)), n / 2)
```

**Pull request merged**